### PR TITLE
Make space optional in amount with denomination

### LINKED ISSUE #1519: Amount with denomination is hard to pass as a command line argument
I'm writing a command line wallet, and I'd like to be able to accept amounts as command line arguments. For clarity, I'd like to require that the user always provides a denomination.

However, `Amount::from_str_with_denomination`, requires that there be a space between the number and the denomination, which would have to be quoted if it were passed as a command line argument, e.g.:

```
wallet 1 sat # error, two arguments
wallet '1 sat' # works, but cumbersome
```

Relaxing `Amount::from_str_with_denomination` to make the space optional would make it possible to pass a denomination on the command line without quoting. Does that sound reasonable?

### PR DESCRIPTION
**Pull request by casey**
I didn't add tests for parsing with no space, but wanted to get a PR up to show the approach.

Fixes #1519.

### CODE REVIEW

**Comment by apoelstra** [comment]
concept ACK from me, but this part of the codebase has seen a huge amount of debate that I did not participate in.

File: bitcoin/src/amount.rs [hunk 1]
```diff
@@ -1899,9 +1896,9 @@ mod tests {
         let p = Amount::from_str;
         let sp = SignedAmount::from_str;
 
-        assert_eq!(p("x BTC"), Err(E::InvalidCharacter('x')));
-        assert_eq!(p("5 BTC BTC"), Err(E::InvalidFormat));
-        assert_eq!(p("5 5 BTC"), Err(E::InvalidFormat));
+        assert_eq!(p("x BTC"), Err(E::UnknownDenomination("x BTC".into())));
+        assert_eq!(p("5 BTC BTC"), Err(E::UnknownDenomination("BTC BTC".into())));
```
Comment (Kixunil) [inline bitcoin/src/amount.rs:1900]: This one is an improvement, nice!

File: bitcoin/src/amount.rs [hunk 2]
```diff
@@ -1899,9 +1896,9 @@ mod tests {
         let p = Amount::from_str;
         let sp = SignedAmount::from_str;
 
-        assert_eq!(p("x BTC"), Err(E::InvalidCharacter('x')));
-        assert_eq!(p("5 BTC BTC"), Err(E::InvalidFormat));
-        assert_eq!(p("5 5 BTC"), Err(E::InvalidFormat));
+        assert_eq!(p("x BTC"), Err(E::UnknownDenomination("x BTC".into())));
```
Comment (Kixunil) [inline bitcoin/src/amount.rs:1899]: Would it make sense to add `rfind(' ')` in the error path to check for this? It'd make a nicer error message.

**Review by Kixunil — COMMENTED** [review]
Concept ACK but I think it'll make things like #1406 more challenging. Should be doable though.

File: bitcoin/src/amount.rs (hunk 2 above)
Comment (casey) [reply bitcoin/src/amount.rs]: To make the error messages a little better, I search for a space first, and only if I don't find one, search for the first alphabetic character.

**Comment by casey** [comment]
I finished up the tests and the implementation for SignedAmount, so this should be ready for review.

File: bitcoin/src/amount.rs [hunk 3]
```diff
@@ -1896,39 +1894,51 @@ mod tests {
...
+        case("18446744073709551616 sat", Err(E::TooBig));
+
+        scase("0 msat", Err(E::TooPrecise));
+        scase("-0 msat", Err(E::TooPrecise));
+        scase("000 msat", Err(E::TooPrecise));
+        scase("-000 msat", Err(E::TooPrecise));
+        case("0 msat", Err(E::TooPrecise));
+        case("-0 msat", Err(E::TooPrecise));
+        case("000 msat", Err(E::TooPrecise));
+        case("-000 msat", Err(E::TooPrecise));
```
Comment (Kixunil) [inline bitcoin/src/amount.rs:1934]: These don't look right, 0 msat is just 0. But I guess the bug was there before?

File: bitcoin/src/amount.rs [hunk 4]
```diff
@@ -1896,39 +1894,51 @@ mod tests {
...
-        assert_eq!(p("0.10000000 BTC"), Ok(Amount::from_sat(100_000_00)));
-        assert_eq!(sp("-100 bits"), Ok(SignedAmount::from_sat(-10_000)));
+
+        assert_eq!(Amount::from_str("x BTC"), Err(E::InvalidCharacter('x')));
+        assert_eq!(Amount::from_str("xBTC"), Err(E::UnknownDenomination("xBTC".into())));
+        assert_eq!(Amount::from_str("5 BTC BTC"), Err(E::UnknownDenomination("BTC BTC".into())));
+        assert_eq!(Amount::from_str("5BTC BTC"), Err(E::InvalidCharacter('B')));
+        assert_eq!(Amount::from_str("5 5 BTC"), Err(E::UnknownDenomination("5 BTC".into())));
+
+        #[track_caller]
```
[the remark this answers is not in the document]
Comment (casey) [reply bitcoin/src/amount.rs]: Thanks!

File: bitcoin/src/amount.rs (hunk 3 above)
Comment (casey) [reply bitcoin/src/amount.rs:1934]: Yah, these looked weird to me. I feel like 0 should parse with any denomination, but they were there before so I left them.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Unfortunately, we require that each commit passes test suite. But you'll need to at least fix clippy issue anyway, so just rebase when you do.

**Comment by casey** [comment]
Ahh, got it. Fixed and squashed.

**Review by apoelstra — APPROVED** [review]
ACK 16c49df688b322cae31e5deb017ec9f5be519d44

**Review by Kixunil — APPROVED** [review]
ACK 16c49df688b322cae31e5deb017ec9f5be519d44

**Pull request merged**